### PR TITLE
Keep cloud agents after configuration

### PR DESCRIPTION
**Pull request by ingwarsw**
- [x] Please describe what you did
Filtering nodes to save, so it does not include cloud ones..
Actually I have no idea if what I have done is right :) Tests shows that it works..
Im not sure why but `getter` was never called.

- [x] Link to issue you're working on if there's a relevant one
- #524 
- #715 

- [x] Did you provide a test-case to demonstrate feature actually works / issue is fixed ?

- [ ] Please also consider adding a line to [CHANGELOG.md](https://github.com/jenkinsci/configuration-as-code-plugin/blob/master/CHANGELOG.md)

### CODE REVIEW

File: plugin/src/main/java/io/jenkins/plugins/casc/core/JenkinsConfigurator.java [hunk 1]
```diff
@@ -56,11 +58,24 @@ protected Jenkins instance(Mapping mapping, ConfigurationContext context) {
                 .setter( noop() ));
 
         // Override "nodes" getter so we don't export Nodes registered by Cloud plugins
-        Attribute.<Jenkins,List<Node>>get(attributes, "nodes").ifPresent(a ->
-            a.getter(j -> j.getNodes().stream()
```
Comment (timja) [inline plugin/src/main/java/io/jenkins/plugins/casc/core/JenkinsConfigurator.java]: can you still export the nodes after this change?
This getter was there to stop the cloud nodes showing up in exports

File: integrations/src/test/java/io/jenkins/plugins/casc/core/CloudSupportTest.java [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+    @ConfiguredWithCode("CloudSupportTest.yml")
+    public void should_keep_cloud_ephemeral_nodes_configured_after_reload() throws Exception {
+        Assert.assertEquals("Base nodes not found", 2, j.jenkins.getNodes().size());
+        Assert.assertNotNull("Slave 1", j.jenkins.getNode("agent1"));
+        Assert.assertNotNull("Slave 1", j.jenkins.getNode("agent2"));
+
+        final Node slave = new Cloud2PretendSlave();
+        j.jenkins.addNode(slave);
+
+        ConfigurationAsCode.get().configure(this.getClass().getResource("CloudSupportTest.yml").toString());
```
Comment (timja) [inline integrations/src/test/java/io/jenkins/plugins/casc/core/CloudSupportTest.java]: you don't need this if you're using the `ConfiguredWithCode` annotation you're

File: integrations/src/test/java/io/jenkins/plugins/casc/core/CloudSupportTest.java [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+        Assert.assertEquals("Cloud nodes not found", 3, j.jenkins.getNodes().size());
+        Assert.assertNotNull("Slave 1", j.jenkins.getNode("agent1"));
+        Assert.assertNotNull("Slave 1", j.jenkins.getNode("agent2"));
+        Assert.assertNotNull("Slave cloud", j.jenkins.getNode("testCloud"));
+    }
+
+    @Test
+    @ConfiguredWithCode("CloudSupportTest.yml")
```
Comment (timja) [inline integrations/src/test/java/io/jenkins/plugins/casc/core/CloudSupportTest.java]: do you want this here? you're running the config again in the test?

File: integrations/src/test/java/io/jenkins/plugins/casc/core/CloudSupportTest.java [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+    public void should_keep_cloud_ephemeral_nodes_configured_after_reload() throws Exception {
+        Assert.assertEquals("Base nodes not found", 2, j.jenkins.getNodes().size());
```
Comment (timja) [inline integrations/src/test/java/io/jenkins/plugins/casc/core/CloudSupportTest.java]: you shouldn't be asserting this in every test? should be in its own separate test

File: integrations/src/test/java/io/jenkins/plugins/casc/core/CloudSupportTest.java [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+        Assert.assertNotNull("Slave 1", j.jenkins.getNode("agent1"));
+        Assert.assertNotNull("Slave 1", j.jenkins.getNode("agent2"));
+
+        final Node slave = new Cloud2PretendSlave();
+        j.jenkins.addNode(slave);
+
+        ConfigurationAsCode.get().configure(this.getClass().getResource("CloudSupportTest.yml").toString());
+        Assert.assertEquals("Cloud nodes not found", 3, j.jenkins.getNodes().size());
```
Comment (timja) [inline integrations/src/test/java/io/jenkins/plugins/casc/core/CloudSupportTest.java]: normal practice is to add a static import for assert* methods, makes it more natural to read

```suggestion
       assertEquals("Cloud nodes not found", 3, j.jenkins.getNodes().size());
```

File: integrations/src/test/java/io/jenkins/plugins/casc/core/CloudSupportTest.java (hunk 3 above)
Comment (ingwarsw) [reply integrations/src/test/java/io/jenkins/plugins/casc/core/CloudSupportTest.java]: Actually I could remove it in some cases, but in perfect case I want it there..

File: integrations/src/test/java/io/jenkins/plugins/casc/core/CloudSupportTest.java (hunk 4 above)
Comment (ingwarsw) [reply integrations/src/test/java/io/jenkins/plugins/casc/core/CloudSupportTest.java]: I have own test for that.. so I can simplify others..

File: integrations/src/test/java/io/jenkins/plugins/casc/core/CloudSupportTest.java (hunk 5 above)
Comment (ingwarsw) [reply integrations/src/test/java/io/jenkins/plugins/casc/core/CloudSupportTest.java]: Yup I forgot that in the end of the day..

File: plugin/src/main/java/io/jenkins/plugins/casc/core/JenkinsConfigurator.java (hunk 1 above)
Comment (ingwarsw) [reply plugin/src/main/java/io/jenkins/plugins/casc/core/JenkinsConfigurator.java]: Aaaaaa.. thats why it was never called.. I will get it back..

File: plugin/src/test/java/io/jenkins/plugins/casc/core/JenkinsConfiguratorCloudSupportTest.java [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+    public void should_remove_normal_nodes_configured_after_reload() throws Exception {
+        final Node slave = new StaticPretendSlave();
+        j.jenkins.addNode(slave);
+
+        ConfigurationAsCode.get().configure(this.getClass().getResource("JenkinsConfiguratorCloudSupportTest.yml").toString());
+        assertEquals("Base nodes not found", 2, j.jenkins.getNodes().size());
+    }
+
+    @Test
+    public void should_keep_cloud_no_instantiable_nodes_configured_after_reload() throws Exception {
```
Comment (timja) [inline plugin/src/test/java/io/jenkins/plugins/casc/core/JenkinsConfiguratorCloudSupportTest.java]: could you add a test for exporting the config to make sure it exports what is expected please?
Something like:
https://github.com/jenkinsci/configuration-as-code-plugin/pull/784/files#diff-63c3227aa81b1bacf4cec3ad0f7274a6R74

File: plugin/src/main/java/io/jenkins/plugins/casc/core/JenkinsConfigurator.java [hunk 7]
```diff
@@ -56,11 +55,22 @@ protected Jenkins instance(Mapping mapping, ConfigurationContext context) {
...
-                    .filter(node -> node.getDescriptor().isInstantiable())
-                    .collect(Collectors.toList())
-            )
+        Attribute.<Jenkins, List<Node>>get(attributes, "nodes").ifPresent(attribute ->
+                attribute
+                        .getter(jenkins -> jenkins.getNodes().stream()
+                                .filter(node -> !isCloudNode(node))
```
Comment (casz) [inline plugin/src/main/java/io/jenkins/plugins/casc/core/JenkinsConfigurator.java]: Just a suggestion :sweat_smile: 
For simplicity, you could have a method `isNotCloudNode` and use a method reference.

File: plugin/src/main/java/io/jenkins/plugins/casc/core/JenkinsConfigurator.java [hunk 8]
```diff
@@ -56,11 +55,22 @@ protected Jenkins instance(Mapping mapping, ConfigurationContext context) {
...
+                                .collect(Collectors.toList()))
+                        .setter((jenkins, configuredNodes) -> {
```
Comment (casz) [inline plugin/src/main/java/io/jenkins/plugins/casc/core/JenkinsConfigurator.java]: Regarding my comment on https://github.com/jenkinsci/configuration-as-code-plugin/issues/715#issuecomment-476883521 This looks like I was right, we were missing a setter :smile:

File: plugin/src/main/java/io/jenkins/plugins/casc/core/JenkinsConfigurator.java (hunk 7 above)
Comment (ingwarsw) [reply plugin/src/main/java/io/jenkins/plugins/casc/core/JenkinsConfigurator.java]: Done before you mentioned it 👍

File: plugin/src/main/java/io/jenkins/plugins/casc/core/JenkinsConfigurator.java (hunk 8 above)
Comment (ingwarsw) [reply plugin/src/main/java/io/jenkins/plugins/casc/core/JenkinsConfigurator.java]: :)

File: plugin/src/test/java/io/jenkins/plugins/casc/core/JenkinsConfiguratorCloudSupportTest.java (hunk 6 above)
Comment (ingwarsw) [reply plugin/src/test/java/io/jenkins/plugins/casc/core/JenkinsConfiguratorCloudSupportTest.java]: Added export test..
Was there any reason not to use spock or some similar tool in tests..
JUnit after using spock for a while is really `bulky` :)

Comment (casz) [reply plugin/src/test/java/io/jenkins/plugins/casc/core/JenkinsConfiguratorCloudSupportTest.java]: I know some plugins use Spock and if it makes sense, I'd welcome it. Can we reuse the test classes that are in `misc` package?

Perhaps best done in a separate PR.

Comment (ingwarsw) [reply plugin/src/test/java/io/jenkins/plugins/casc/core/JenkinsConfiguratorCloudSupportTest.java]: I think that migration right now will not be that strait forward.. 
`spock` => `groovy` :)

Things from `misc` would work as a charm...

Comment (ingwarsw) [reply plugin/src/test/java/io/jenkins/plugins/casc/core/JenkinsConfiguratorCloudSupportTest.java]: And for sure it needs extra PR, if ever :)

Comment (casz) [reply plugin/src/test/java/io/jenkins/plugins/casc/core/JenkinsConfiguratorCloudSupportTest.java]: I don't mind groovy, we live and breath it everyday in Jenkins pipelines :sweat_smile:

File: plugin/src/main/java/io/jenkins/plugins/casc/core/JenkinsConfigurator.java (hunk 7 above)
Comment (timja) [reply plugin/src/main/java/io/jenkins/plugins/casc/core/JenkinsConfigurator.java]: not done?
Casz meant something like:
```suggestion
                                .filter(this::isNotCloudNode)
```

Comment (ingwarsw) [reply plugin/src/main/java/io/jenkins/plugins/casc/core/JenkinsConfigurator.java]: Once I need `isCloudNode` and once `isNotCloudNode`

**Comment by ingwarsw** [comment]
@timja we have found one extra issue.. see https://github.com/jenkinsci/google-compute-engine-plugin/issues/53#issuecomment-478080909

File: plugin/src/main/java/io/jenkins/plugins/casc/core/JenkinsConfigurator.java (hunk 7 above)
Comment (casz) [reply plugin/src/main/java/io/jenkins/plugins/casc/core/JenkinsConfigurator.java]: Yes sorry for not using suggestions

Comment (casz) [reply plugin/src/main/java/io/jenkins/plugins/casc/core/JenkinsConfigurator.java]: `isNotCloudNode` will `return !isCloudNode()`

**Comment by casz** [comment]
:raised_hands: for Travis's speed :smile: finished before Codacy and Jenkins

image

**Comment by timja** [comment]
Tests are failing:
```
io.jenkins.plugins.casc.core.JenkinsConfiguratorCloudSupportTest.should_export_only_static_nodes(io.jenkins.plugins.casc.core.JenkinsConfiguratorCloudSupportTest)
[ERROR]   Run 1: JenkinsConfiguratorCloudSupportTest.should_export_only_static_nodes:96->toYamlString:120 » NullPointer
[ERROR]   Run 2: JenkinsConfiguratorCloudSupportTest.should_export_only_static_nodes:96->toYamlString:120 » NullPointer
[ERROR]   Run 3: JenkinsConfiguratorCloudSupportTest.should_export_only_static_nodes:96->toYamlString:120 » NullPointer
[ERROR]   Run 4: JenkinsConfiguratorCloudSupportTest.should_export_only_static_nodes:96->toYamlString:120 » NullPointer
[ERROR]   Run 5: JenkinsConfiguratorCloudSupportTest.should_export_only_static_nodes:96->toYamlString:120 » NullPointer
```

**Comment by ingwarsw** [comment]
@timja Yes I see.. after merging develop.. I will check it..

**Pull request merged**